openjdk/mmtkUpcalls.cpp

```diff
@@ -120,12 +124,25 @@ static void mmtk_spawn_collector_thread(void* tls, int kind, void* ctx) {
 static void mmtk_block_for_gc() {
   MMTkHeap::heap()->_last_gc_time = os::javaTimeNanos() / NANOSECS_PER_MILLISEC;
   log_debug(gc)("Thread (id=%d) will block waiting for GC to finish.", Thread::current()->osthread()->thread_id());
+
+  // We must read the counter before entering safepoint.
+  // This thread has just triggered GC.
+  // Before this thread enters safe point, the GC cannot start, and therefore cannot finish,
+  // and cannot increment the counter mmtk_start_the_world_count.
+  // Otherwise, if we attempt to acquire the gc_lock first, GC may have triggered stop-the-world
+  // first, and this thread will be blocked for the entire stop-the-world duration before it can
+  // get the lock.  Once that happens, the current thread will read the counter after the GC, and
+  // wait for the next non-existing GC forever.
+  size_t my_count = Atomic::load(&mmtk_start_the_world_count);
```

Comment:
The following is just speculation. Let me know if I am wrong.

---

`block_for_gc()` does not mean this thread triggered GC -- it means the thread tries to allocate and MMTk cannot fulfill the allocation request and need the thread to be blocked until a GC is done. 

So `block_for_gc()` may be called multiple times for a GC. The first thread that polls and triggers the GC will be blocked by `block_for_gc()`, but any other thread may also be blocked by `block_for_gc()` if `stop_all_mutators()` does not yet stop that thread, and that thread tries to slow-path allocate. So it is possible that `mmtk_start_the_world_count` is increased multiple times per GC, and that means a thread may be blocked for multiple GCs.

For example, if we have 3 mutator threads, A, B and C, in the worst case:
* thread A in `block_for_gc()`: count = 1
* thread B in `block_for_gc()`: count = 2
* thread C in `block_for_gc()`: count = 3
* GC#1 happens. A resumes, but B and C is still blocked
* thread A in `block_for_gc()`: count = 4
* GC#2 happens. B resumes, but A and C is still blocked.
* ...

Fortunately in this worst case, we will always have 1 thread to resume after each GC. So it won't hang. But this probably will affect performance.

Reply:
Why not do something like a semaphore where after GC#1, the count is reduced by 1 (or reset to 0)?

Reply:
@k-sareen Maybe, but the idea that mutators can (or need to) modify the synchronising object (the counter, the semaphore, etc.) makes me feel unease, just like the case of "counter incremented multiple times in a GC" that Yi mentioned.